src/json.ts

```diff
@@ -0,0 +1,139 @@
+import deepEqual from 'fast-deep-equal';
+import { hasProperty } from './misc';
+
+/**
+ * Any JSON-compatible value.
+ */
+export type Json =
+  | null
+  | boolean
+  | number
+  | string
+  | Json[]
+  | { [prop: string]: Json };
+
+/**
+ * Type guard for {@link Json}.
+ *
+ * @param value - The value to check.
+ * @returns Whether the value is valid JSON.
+ */
+export function isValidJson(value: unknown): value is Json {
+  try {
+    return deepEqual(value, JSON.parse(JSON.stringify(value)));
+  } catch (_) {
+    return false;
+  }
+}
+
+/**
+ * A String specifying the version of the JSON-RPC protocol.
+ * MUST be exactly "2.0".
+ */
+export type JsonRpcVersion2 = '2.0';
+
+/**
+ * The string '2.0'.
+ */
+export const jsonrpc2 = '2.0' as const;
+
+/**
+ * An identifier established by the Client that MUST contain a String, Number,
+ * or NULL value if included. If it is not included it is assumed to be a
+ * notification. The value SHOULD normally not be Null and Numbers SHOULD
+ * NOT contain fractional parts.
+ */
+export type JsonRpcId = number | string | null;
+
+/**
+ * A JSON-RPC error object.
+ */
+export type JsonRpcError = {
+  code: number;
+  message: string;
+  data?: unknown;
+  stack?: string;
+};
+
+/**
+ * A JSON-RPC request object.
+ */
+export type JsonRpcRequest<T> = {
+  jsonrpc: JsonRpcVersion2;
+  method: string;
+  id: JsonRpcId;
+  params?: T;
+};
+
+/**
+ * A JSON-RPC notification object.
+ */
+export type JsonRpcNotification<T> = {
+  jsonrpc: JsonRpcVersion2;
+  method: string;
+  params?: T;
+};
+
+/**
+ * The internal, base type for JSON-RPC responses.
+ */
+type JsonRpcResponseBase = {
+  jsonrpc: JsonRpcVersion2;
+  id: JsonRpcId;
+};
+
+/**
+ * A successful JSON-RPC response object.
+ *
+ * @template T - The type of the result value.
+ */
+export type JsonRpcSuccess<T = unknown> = JsonRpcResponseBase & {
+  result: T;
+};
```

Comment:
Not sure what our conventions are, but for readability this might be better:
```suggestion
export type JsonRpcSuccess<Result = unknown> = JsonRpcResponseBase & {
  result: Result;
};
```
This goes for some other types in this file too.